tsfc/driver.py

```diff
@@ -243,3 +250,30 @@ def facetarea():
 
     kernel_name = "%s_%s_integral_%s" % (prefix, integral_type, integral_data.subdomain_id)
     return builder.construct_kernel(kernel_name, body)
+
+
+def lower_integral_type(fiat_cell, integral_type):
+    dim = fiat_cell.get_dimension()
+    if integral_type == 'cell':
+        integration_dim = dim
+    elif integral_type in ['exterior_facet', 'interior_facet']:
+        integration_dim = dim - 1
+    else:
+        basedim, extrdim = dim
+        assert extrdim == 1
+
+        if integral_type in ['exterior_facet_vert', 'interior_facet_vert']:
+            integration_dim = (basedim - 1, 1)
+        elif integral_type in ['exterior_facet_bottom', 'exterior_facet_top', 'interior_facet_horiz']:
+            integration_dim = (basedim, 0)
+        else:
+            raise NotImplementedError("integral type %s not supported" % integral_type)
+
+    if integral_type == 'exterior_facet_bottom':
+        entity_ids = [0]
+    elif integral_type == 'exterior_facet_top':
+        entity_ids = [1]
+    else:
+        entity_ids = range(len(fiat_cell.get_topology()[integration_dim]))
+
+    return integration_dim, entity_ids
```

Comment:
So what really happens is this:

    (cell, integral_type) -> (cell, integration_dim, entity_ids)

This function doesn't return the cell, because that doesn't change, just decomposes `integral_type` into `(integration_dim, entity_ids)`, but it needs the cell to do that decomposition. The cell still maintains the distinction between a triangle and a quadrilateral. Basically, integration dimension and entity IDs are only used to pass them to methods of the FIAT cells.